File: backend/app/api/v1/store/bot_sessions.py
"""Meeting bot sessions REST API (worker-owned SQLite, app-layer scoping)."""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.v1.store._common import get_db, get_principal, scoped_deal_or_404
from app.db.models import MeetingBotSession
from app.db.scope import Principal, org_scoped
from app.schemas.common import BaseSchema

router = APIRouter()


# ---- schemas --------------------------------------------------------------
class BotSessionCreate(BaseSchema):
    deal_id: str
    platform: str
    meeting_url: str
    scheduled_start: datetime | None = None
    consent_obtained: bool = False


class BotSessionResponse(BaseSchema):
    id: str
    org_id: str
    deal_id: str
    meeting_id: str | None = None
    platform: str
    meeting_url: str
    status: str
    scheduled_start: datetime | None = None
    actual_start: datetime | None = None
    actual_end: datetime | None = None
    recording_file_key: str | None = None
    recall_bot_id: str | None = None
    live_transcript_channel: str | None = None
    consent_obtained: bool
    created_by: str
    created_at: datetime
    updated_at: datetime


# ---- routes ---------------------------------------------------------------
@router.get("/bot-sessions", response_model=list[BotSessionResponse])
def list_bot_sessions(
    deal_id: str | None = Query(None),
    status_filter: str | None = Query(None, alias="status"),
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[BotSessionResponse]:
    stmt = org_scoped(select(MeetingBotSession), MeetingBotSession, principal)
    if deal_id:
        stmt = stmt.where(MeetingBotSession.deal_id == deal_id)
    if status_filter:
        stmt = stmt.where(MeetingBotSession.status == status_filter)
    stmt = stmt.order_by(MeetingBotSession.created_at.desc())
    rows = session.scalars(stmt).all()
    return [BotSessionResponse.model_validate(r) for r in rows]


@router.post(
    "/bot-sessions",
    response_model=BotSessionResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_bot_session(
    payload: BotSessionCreate,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> BotSessionResponse:
    deal = scoped_deal_or_404(session, principal, payload.deal_id)
    bot_session = MeetingBotSession(
        org_id=deal.org_id,
        deal_id=deal.id,
        platform=payload.platform,
        meeting_url=payload.meeting_url,
        status="scheduled",
        scheduled_start=(
            payload.scheduled_start.isoformat() if payload.scheduled_start else None
        ),
        consent_obtained=payload.consent_obtained,
        created_by=principal.user_id,
    )
    session.add(bot_session)
    session.flush()
    return BotSessionResponse.model_validate(bot_session)


@router.post("/bot-sessions/{session_id}/cancel", response_model=BotSessionResponse)
def cancel_bot_session(
    session_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> BotSessionResponse:
    bot_session = session.get(MeetingBotSession, session_id)
    if bot_session is None or not principal.in_org(bot_session.org_id):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Bot session not found"
        )
    bot_session.status = "cancelled"
    session.flush()
    return BotSessionResponse.model_validate(bot_session)

File: worker/internal/store/botsessions.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"strings"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/model"
	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/util"
)

const botSessionCols = "id, org_id, deal_id, meeting_id, platform, meeting_url, status, " +
	"scheduled_start, actual_start, actual_end, recording_file_key, recall_bot_id, " +
	"live_transcript_channel, consent_obtained, created_by, created_at, updated_at"

func scanBotSession(row interface{ Scan(...any) error }) (*model.MeetingBotSession, error) {
	var b model.MeetingBotSession
	err := row.Scan(&b.ID, &b.OrgID, &b.DealID, &b.MeetingID, &b.Platform, &b.MeetingURL,
		&b.Status, &b.ScheduledStart, &b.ActualStart, &b.ActualEnd, &b.RecordingFileKey,
		&b.RecallBotID, &b.LiveTranscriptChannel, &b.ConsentObtained, &b.CreatedBy,
		&b.CreatedAt, &b.UpdatedAt)
	return &b, err
}

// BotSessionFilters are the list query params (ports list_bot_sessions).
type BotSessionFilters struct {
	DealID string
	Status string
}

// ListBotSessions returns the principal's org-scoped bot sessions newest-first
// (ports list_bot_sessions). Every row is constrained to the caller's orgs.
func ListBotSessions(ctx context.Context, conn *sql.DB, p *Principal, f BotSessionFilters) ([]model.MeetingBotSession, error) {
	pred, args := p.OrgFilter("org_id")
	where := []string{pred}
	if f.DealID != "" {
		where = append(where, "deal_id = ?")
		args = append(args, f.DealID)
	}
	if f.Status != "" {
		where = append(where, "status = ?")
		args = append(args, f.Status)
	}
	q := "SELECT " + botSessionCols + " FROM meeting_bot_sessions WHERE " +
		strings.Join(where, " AND ") + " ORDER BY created_at DESC"
	rows, err := conn.QueryContext(ctx, q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var items []model.MeetingBotSession
	for rows.Next() {
		b, err := scanBotSession(rows)
		if err != nil {
			return nil, err
		}
		items = append(items, *b)
	}
	return items, rows.Err()
}

// BotSessionCreate is the create payload.
type BotSessionCreate struct {
	DealID          string
	Platform        string
	MeetingURL      string
	ScheduledStart  *string
	ConsentObtained bool
}

// CreateBotSession inserts a "scheduled" bot session for a deal the principal can
// see (ports create_bot_session). The deal must be in one of the caller's orgs,
// else ErrNotFound — the new row inherits the deal's org_id (never the caller's).
//
// TODO(recall-phase): the Python worker stops at the DB write here; dispatching
// the Recall.ai bot happens out-of-band. When the Recall integration is ported,
// kick off the bot here and persist recall_bot_id / live_transcript_channel.
func CreateBotSession(ctx context.Context, conn *sql.DB, p *Principal, in BotSessionCreate) (*model.MeetingBotSession, error) {
	deal, err := ScopedDeal(ctx, conn, p, in.DealID)
	if err != nil {
		return nil, err
	}
	now, id := util.NowISO(), util.NewUUID()
	if _, err := conn.ExecContext(ctx,
		"INSERT INTO meeting_bot_sessions("+botSessionCols+") VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
		id, deal.OrgID, deal.ID, nil, in.Platform, in.MeetingURL, "scheduled",
		in.ScheduledStart, nil, nil, nil, nil, nil, in.ConsentObtained, p.UserID, now, now,
	); err != nil {
		return nil, err
	}
	return getBotSession(ctx, conn, p, id)
}

// getBotSession fetches an org-scoped bot session by id, else ErrNotFound.
func getBotSession(ctx context.Context, conn *sql.DB, p *Principal, id string) (*model.MeetingBotSession, error) {
	pred, args := p.OrgFilter("org_id")
	q := "SELECT " + botSessionCols + " FROM meeting_bot_sessions WHERE id = ? AND " + pred
	b, err := scanBotSession(conn.QueryRowContext(ctx, q, append([]any{id}, args...)...))
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrNotFound
	}
	return b, err
}

// CancelBotSession marks a bot session "cancelled" (ports cancel_bot_session).
// The row must be in one of the caller's orgs, else ErrNotFound (a foreign row is
// indistinguishable from a missing one).
//
// TODO(recall-phase): the Python worker only flips the status; when the Recall
// integration lands, also signal Recall.ai to stop/leave the live bot here.
func CancelBotSession(ctx context.Context, conn *sql.DB, p *Principal, id string) (*model.MeetingBotSession, error) {
	if _, err := getBotSession(ctx, conn, p, id); err != nil {
		return nil, err
	}
	if _, err := conn.ExecContext(ctx,
		"UPDATE meeting_bot_sessions SET status = ?, updated_at = ? WHERE id = ?",
		"cancelled", util.NowISO(), id); err != nil {
		return nil, err
	}
	return getBotSession(ctx, conn, p, id)
}
